### ISSUE TITLE
Using as-pect for benchmarking

### ISSUE THREAD

**Issue by jedisct1** [jtenner/as-pect#24]
As-pect currently reports how long each test took to run. However, a single iteration, especially on tests that quickly complete, doesn't provide very accurate results.

I trust you to find the best way to expose this, but here are some of the things the benchmark mode could do:

- Run multiple iterations of every test until the test has run for a minimum amount of time.
- Or, if that can be configurable per-test, run a configurable number of iterations (useful for tests that take a long time to complete).
- Keep the measurements for each iteration, so that we can output the median, min, max and variance. The variance is useful to see if the results are reliable or if the host was busy doing something else while the tests were running.

There are probably many improvements that could follow, but having only these would already make as-pect very nice for benchmarking purposes.

**Comment by jtenner** [comment]
This testing library was inspired by `jest` and it doesn't look like there is a lot of buzz on the internet about benchmarking with jest. I found an issue involving `benchmark.js` which appears to still be open. However, it is mostly unrelated.

Running tests in a benchmarked way could simply be the default method of testing here.

Some options:

# A Benchmark Test

This is only a small example of what a `benchmark` *could* look like.

```ts
benchmark(benchName,  iterationCount, (): void => {
  // run code here
});
```
I don't like how to this solution feels. I think benchmarking should be a cornerstone of the `TestContext` class located in `./src/test/TestContext.ts` file.

# Enabled via Command line or configuration

```json
{
  "benchmark": {
    "maxIterations": 100, // default
    "maxTime": 100, //ms default
    "minIterations: 5 //default
  }
}
```
These options could also be implemented via command line, and overwrite the configuration file.

# Via Script Transformer

This would be the most work-intensive solution I could think of, but the creators of AssemblyScript seem to think that using decorators is possible and would have to be transformed when the command line runs.

```ts
it("should run things", @benchmark (): void => {
  // test here
});
```

I don't know what it would have to transform into, but this would effectively mark the exported function as a benchmarkable function which would collect statistics.

Any thoughts?

**Comment by jedisct1** [comment]
Using a simple configuration file would be fine, but maybe with more granularity: a "default" set of properties, that can be overridden for specific tests.

**Comment by jtenner** [comment]
That's doable. How would test overrides look?

**Comment by jtenner** [comment]
Please see #26 for the massive pull request involving performance metrics.

**Comment by jedisct1** [comment]
Awesome work.  I'm gonna give it a try right away.

**Comment by jtenner** [comment]
closed via #26
